Fix gesture conflicts of context_menu

xhzq233 opened the pull request:
Fixes #70716, #81057
Related ##52226

## Why the conflict

If we put a `TapGestureRecognizer` at `CupertinoContextMenu` 's subtree. 

For simplicity, we name the `TapGestureRecognizer` inside `CupertinoContextMenu` as **tg2**, the `TapGestureRecognizer` at `CupertinoContextMenu` 's subtree as **tg1**, and the press duration as **t1**.

When `kPressTimeout+_previewLongPressTimeout/2` (500ms) < **t1** < `kPressTimeout+_previewLongPressTimeout `(900ms), this moment

1. Although **tg2** was rejected, `_ContextMenuRoute` was still opened.

2.  And the onTap method of **tg1** is triggered

The problem is that the GestureRecognizer in this ContextMenu is TapGestureRecognizer instead of LongPressGestureRecognizer, and Tap does not claim to victory, but GestureArena selects the deepest Tap to win.

1. When **t1** is greater than 500ms, it means that it has passed `PrimaryPointerGestureRecognizer.deadline` plus half of the `openController`'s duration, means that `_ContextMenuRoute` is about to be opened. After `GestureArena` received pointer up event, **tg1** competed with **tg2**. The default "child-victory" worked, and the onTap of tg1 was called, so both were triggered.

2. When **t1** is less than 900ms, because route is opened at this time, and both will be rejected by arena.

## Solution

The reason is that `TapGestureRecognizer` will not declare victory by itself. Add the following code.

```dart

// call this when animation's value first reaches [_midpoint]

_tapGestureRecognizer.resolve(GestureDisposition.accepted);

```


## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

xhzq233 committed:
Fix gesture conflicts of context_menu

xhzq233 committed:
Fix gesture conflicts of context_menu

engine-flutter-autoroll committed:
Roll Flutter Engine from 2df3b9c4b2a4 to e40995da7869 (2 revisions) (#130985)

https://github.com/flutter/engine/compare/2df3b9c4b2a4...e40995da7869

2023-07-20 skia-flutter-autoroll@skia.org Roll ANGLE from 5e38a31bd76a to a4c283be741f (1 revision) (flutter/engine#43849)
2023-07-20 skia-flutter-autoroll@skia.org Roll Skia from 65a83c4de7f2 to 401c85ab1e21 (1 revision) (flutter/engine#43848)

If this roll has caused a breakage, revert this CL and stop the roller
using the controls here:
https://autoroll.skia.org/r/flutter-engine-flutter-autoroll
Please CC bdero@google.com,rmistry@google.com,zra@google.com on the revert to ensure that a human
is aware of the problem.

To file a bug in Flutter: https://github.com/flutter/flutter/issues/new/choose

To report a problem with the AutoRoller itself, please file a bug:
https://bugs.chromium.org/p/skia/issues/entry?template=Autoroller+Bug

Documentation for the AutoRoller is here:
https://skia.googlesource.com/buildbot/+doc/main/autoroll/README.md

polina-c committed:
Upgrade leak_tracker. (#130951)

engine-flutter-autoroll committed:
Roll Flutter Engine from e40995da7869 to c645eb6da8a9 (1 revision) (#130988)

https://github.com/flutter/engine/compare/e40995da7869...c645eb6da8a9

2023-07-20 skia-flutter-autoroll@skia.org Roll Fuchsia Linux SDK from ZwCUlo28olVzLlqTl... to hVAAd2NqYOjUF_I99... (flutter/engine#43850)

Also rolling transitive DEPS:
  fuchsia/sdk/core/linux-amd64 from ZwCUlo28olVz to hVAAd2NqYOjU

If this roll has caused a breakage, revert this CL and stop the roller
using the controls here:
https://autoroll.skia.org/r/flutter-engine-flutter-autoroll
Please CC bdero@google.com,rmistry@google.com,zra@google.com on the revert to ensure that a human
is aware of the problem.

To file a bug in Flutter: https://github.com/flutter/flutter/issues/new/choose

To report a problem with the AutoRoller itself, please file a bug:
https://bugs.chromium.org/p/skia/issues/entry?template=Autoroller+Bug

Documentation for the AutoRoller is here:
https://skia.googlesource.com/buildbot/+doc/main/autoroll/README.md

engine-flutter-autoroll committed:
Roll Packages from 209db219ad73 to 674179f97bda (4 revisions) (#130989)

https://github.com/flutter/packages/compare/209db219ad73...674179f97bda

2023-07-19 stuartmorgan@google.com [image_picker] Deprecate platform interface methods (flutter/packages#4520)
2023-07-19 47866232+chunhtai@users.noreply.github.com [go_router] Adds a parentNavigatorKey parameter to ShellRouteData. (flutter/packages#4409)
2023-07-19 stuartmorgan@google.com [ci] Shard Windows Dart unit tests (flutter/packages#4519)
2023-07-19 engine-flutter-autoroll@skia.org Roll Flutter from 6f09064e785b to d07e8aece184 (60 revisions) (flutter/packages#4522)

If this roll has caused a breakage, revert this CL and stop the roller
using the controls here:
https://autoroll.skia.org/r/flutter-packages-flutter-autoroll
Please CC flutter-ecosystem@google.com,rmistry@google.com on the revert to ensure that a human
is aware of the problem.

To file a bug in Flutter: https://github.com/flutter/flutter/issues/new/choose

To report a problem with the AutoRoller itself, please file a bug:
https://bugs.chromium.org/p/skia/issues/entry?template=Autoroller+Bug

Documentation for the AutoRoller is here:
https://skia.googlesource.com/buildbot/+doc/main/autoroll/README.md

bleroux committed:
Update about tests for M3 (#130970)

This PR updates unit tests in `about_test.dart` for M3 migration.

More info in https://github.com/flutter/flutter/issues/127064

- Two tests were failing in M3 due to a memory leak. As the memory leak is now fixed, see https://github.com/flutter/flutter/pull/130720, this two tests does not depend anymore on the Material version.

- Created several M3 tests related to typography and rendering changes.

bleroux committed:
Update AutoComplete test for M3 migration (#130883)

This PR updates unit tests from autocomplete_test.dart for M3 migration.

More info in https://github.com/flutter/flutter/issues/127064

I replaced magic numbers (64.0 and 187.0) and found ways to compute them without relying on the Material version.

engine-flutter-autoroll committed:
Roll Flutter Engine from c645eb6da8a9 to 6d7842d25f81 (1 revision) (#130992)

https://github.com/flutter/engine/compare/c645eb6da8a9...6d7842d25f81

2023-07-20 skia-flutter-autoroll@skia.org Roll Dart SDK from 603aacd8400f to 857c9a2ae14a (1 revision) (flutter/engine#43851)

If this roll has caused a breakage, revert this CL and stop the roller
using the controls here:
https://autoroll.skia.org/r/flutter-engine-flutter-autoroll
Please CC bdero@google.com,rmistry@google.com,zra@google.com on the revert to ensure that a human
is aware of the problem.

To file a bug in Flutter: https://github.com/flutter/flutter/issues/new/choose

To report a problem with the AutoRoller itself, please file a bug:
https://bugs.chromium.org/p/skia/issues/entry?template=Autoroller+Bug

Documentation for the AutoRoller is here:
https://skia.googlesource.com/buildbot/+doc/main/autoroll/README.md

QuncCccccc committed:
Update `TextSelectionTheme`, `ThemeData`, `TimePicker`, and `TimePickerTheme` tests for M2/M3 (#130547)

Updated unit tests for `TextSelectionTheme`, `ThemeData`, `TimePicker` and `TimePickerTheme` to have M2 and M3 versions.

More info in #127064

MitchellGoodwin commented:
Thank you for your contribution! This will need to have some kind of test on it in the code. Would it be possible to write at least one that watches to see if the different tap events fire after press and hold?

engine-flutter-autoroll committed:
Roll Flutter Engine from 6d7842d25f81 to a3fc18514cd6 (2 revisions) (#131007)

https://github.com/flutter/engine/compare/6d7842d25f81...a3fc18514cd6

2023-07-20 skia-flutter-autoroll@skia.org Roll Skia from b8133dda3a8c to a3aca7ae523e (1 revision) (flutter/engine#43855)
2023-07-20 skia-flutter-autoroll@skia.org Roll Skia from 401c85ab1e21 to b8133dda3a8c (4 revisions) (flutter/engine#43853)

If this roll has caused a breakage, revert this CL and stop the roller
using the controls here:
https://autoroll.skia.org/r/flutter-engine-flutter-autoroll
Please CC bdero@google.com,rmistry@google.com,zra@google.com on the revert to ensure that a human
is aware of the problem.

To file a bug in Flutter: https://github.com/flutter/flutter/issues/new/choose

To report a problem with the AutoRoller itself, please file a bug:
https://bugs.chromium.org/p/skia/issues/entry?template=Autoroller+Bug

Documentation for the AutoRoller is here:
https://skia.googlesource.com/buildbot/+doc/main/autoroll/README.md

engine-flutter-autoroll committed:
Roll Flutter Engine from a3fc18514cd6 to 062079ba30b6 (3 revisions) (#131010)

https://github.com/flutter/engine/compare/a3fc18514cd6...062079ba30b6

2023-07-20 dnfield@google.com More validation logs for CommandEncoderVK submission (flutter/engine#43859)
2023-07-20 skia-flutter-autoroll@skia.org Roll Dart SDK from 857c9a2ae14a to 1df95f328d0c (1 revision) (flutter/engine#43858)
2023-07-20 mdebbar@google.com [web] Preserve correct CanvasKit Variant during test initialization (flutter/engine#43854)

If this roll has caused a breakage, revert this CL and stop the roller
using the controls here:
https://autoroll.skia.org/r/flutter-engine-flutter-autoroll
Please CC bdero@google.com,rmistry@google.com,zra@google.com on the revert to ensure that a human
is aware of the problem.

To file a bug in Flutter: https://github.com/flutter/flutter/issues/new/choose

To report a problem with the AutoRoller itself, please file a bug:
https://bugs.chromium.org/p/skia/issues/entry?template=Autoroller+Bug

Documentation for the AutoRoller is here:
https://skia.googlesource.com/buildbot/+doc/main/autoroll/README.md

HansMuller committed:
Updated the ThemeData API example (#130954)

Hixie committed:
Trivial grammar and wrapping fix for docs (#130955)

engine-flutter-autoroll committed:
Roll Flutter Engine from 062079ba30b6 to 9b2ebf2afe00 (2 revisions) (#131013)

https://github.com/flutter/engine/compare/062079ba30b6...9b2ebf2afe00

2023-07-20 skia-flutter-autoroll@skia.org Roll Skia from a3aca7ae523e to 18e834916f47 (1 revision) (flutter/engine#43860)
2023-07-20 mdebbar@google.com Add url to get GoogleFonts API key (flutter/engine#43857)

If this roll has caused a breakage, revert this CL and stop the roller
using the controls here:
https://autoroll.skia.org/r/flutter-engine-flutter-autoroll
Please CC bdero@google.com,rmistry@google.com,zra@google.com on the revert to ensure that a human
is aware of the problem.

To file a bug in Flutter: https://github.com/flutter/flutter/issues/new/choose

To report a problem with the AutoRoller itself, please file a bug:
https://bugs.chromium.org/p/skia/issues/entry?template=Autoroller+Bug

Documentation for the AutoRoller is here:
https://skia.googlesource.com/buildbot/+doc/main/autoroll/README.md

christopherfujino committed:
Migrate more integration tests to process result matcher (#130994)

Part of https://github.com/flutter/flutter/issues/127135

gspencergoog committed:
Add applyFocusChangeIfNeeded, have menus restore focus before activating (#130536)

## Description

This modifies the `MenuAnchor` `onPressed` activation to delay until after the current frame is built, and resolve any focus changes before it invokes the `onPressed`, so that actions that operate on the `primaryFocus` can have a chance of working on the focused item they were meant to work on.

## Related Issues
 - Fixes https://github.com/flutter/flutter/issues/118731

## Tests
 - No tests yet (hence draft still)

engine-flutter-autoroll committed:
Roll Flutter Engine from 9b2ebf2afe00 to ab7d424d4467 (4 revisions) (#131015)

https://github.com/flutter/engine/compare/9b2ebf2afe00...ab7d424d4467

2023-07-20 jonahwilliams@google.com [Impeller] Provide fragment uniform data through varyings for solid color, glyph atlas, texture shaders. (flutter/engine#43838)
2023-07-20 skia-flutter-autoroll@skia.org Roll ANGLE from a4c283be741f to f2e0f8a0b236 (2 revisions) (flutter/engine#43864)
2023-07-20 skia-flutter-autoroll@skia.org Roll Skia from 18e834916f47 to 100d0f858f02 (7 revisions) (flutter/engine#43863)
2023-07-20 skia-flutter-autoroll@skia.org Roll Fuchsia Mac SDK from SmAtKPfGzPllC9gfO... to -SaPL-46jpiYbnCAu... (flutter/engine#43862)

Also rolling transitive DEPS:
  fuchsia/sdk/core/mac-amd64 from SmAtKPfGzPll to -SaPL-46jpiY

If this roll has caused a breakage, revert this CL and stop the roller
using the controls here:
https://autoroll.skia.org/r/flutter-engine-flutter-autoroll
Please CC bdero@google.com,rmistry@google.com,zra@google.com on the revert to ensure that a human
is aware of the problem.

To file a bug in Flutter: https://github.com/flutter/flutter/issues/new/choose

To report a problem with the AutoRoller itself, please file a bug:
https://bugs.chromium.org/p/skia/issues/entry?template=Autoroller+Bug

Documentation for the AutoRoller is here:
https://skia.googlesource.com/buildbot/+doc/main/autoroll/README.md

flutter-pub-roller-bot committed:
Roll pub packages (#131022)

This PR was generated by `flutter update-packages --force-upgrade`.

stuartmorgan-g committed:
Use downgraded analyze for flutter/packages (#130878)

Add the `--downgrade` flag to the flutter/packages analyze test, so that it won't get the latest version of dependencies, avoiding out-of-band failures in flutter/flutter due to package publishing events.

Fixes https://github.com/flutter/flutter/issues/129633

engine-flutter-autoroll committed:
Roll Flutter Engine from ab7d424d4467 to 218b71cd7a45 (2 revisions) (#131024)

https://github.com/flutter/engine/compare/ab7d424d4467...218b71cd7a45

2023-07-20 skia-flutter-autoroll@skia.org Roll Skia from 981146e6305d to 6f2b2e94ebbd (3 revisions) (flutter/engine#43871)
2023-07-20 skia-flutter-autoroll@skia.org Roll Skia from 100d0f858f02 to 981146e6305d (3 revisions) (flutter/engine#43866)

If this roll has caused a breakage, revert this CL and stop the roller
using the controls here:
https://autoroll.skia.org/r/flutter-engine-flutter-autoroll
Please CC bdero@google.com,rmistry@google.com,zra@google.com on the revert to ensure that a human
is aware of the problem.

To file a bug in Flutter: https://github.com/flutter/flutter/issues/new/choose

To report a problem with the AutoRoller itself, please file a bug:
https://bugs.chromium.org/p/skia/issues/entry?template=Autoroller+Bug

Documentation for the AutoRoller is here:
https://skia.googlesource.com/buildbot/+doc/main/autoroll/README.md

xhzq233 committed:
Merge branch 'cupertino_ctx_menu_improvement' of https://github.com/xhzq233/flutter into cupertino_ctx_menu_improvement

xhzq233 committed:
Added test to check whether the gesture conflict is resolved.

xhzq233 commented:
Sorry that I merged the master by mistake. Let me open another pr.

xhzq233 commented:
new pr #131030.
